wom/src/main/scala/wdl4s/wom/graph/ConditionalNode.scala

```diff
@@ -0,0 +1,63 @@
+package wdl4s.wom.graph
+
+import cats.data.Validated.Valid
+import cats.syntax.validated._
+import cats.syntax.apply._
+import lenthall.validation.ErrorOr.ErrorOr
+import lenthall.validation.ErrorOr.ShortCircuitingFlatMap
+import wdl4s.wom.graph.GraphNodePort.{ConditionalOutputPort, InputPort, OutputPort}
+import wdl4s.wdl.types.{WdlBooleanType, WdlOptionalType}
+import wdl4s.wom.graph.GraphNode.LinkedInputPort
+
+/**
+  * Currently only WDL has the concept of conditional executions:
+  *
+  * @param innerGraph Imagine that the contents of the conditional block were a self-contained workflow. That's this Graph
+  * @param condition The (boolean) expression on which the conditional is predicated.
+  * @param innerGraphInputs The set of input ports that will be needed to satisfy the inner graph's inputs.
+  * @param outputMapping Output ports for the scatter node, which also link back to GraphOutputNodes of the inner graph.
+  */
+final case class ConditionalNode private(innerGraph: Graph,
+                                         condition: InstantiatedExpression,
+                                         innerGraphInputs: Set[InputPort],
+                                         outputMapping: Set[ConditionalOutputPort]) extends GraphNode {
+
+  override val name: String = "ConditionalNode"
+
+  // NB if you find yourself calling .filter on this set of inputPorts, you probably just wanted to access either
+  // the scatterVariableMapping or otherInputPorts fields directly.
+  override val inputPorts: Set[InputPort] = condition.inputPorts ++ innerGraphInputs
+  override val outputPorts: Set[GraphNodePort.OutputPort] = outputMapping.toSet[OutputPort]
+}
+
+object ConditionalNode  {
+
+  final case class ConditionalNodeWithInputs(conditionalNode: ConditionalNode, inputs: Set[GraphInputNode])
+
+  def wireInConditional(innerGraph: Graph, condition: GraphNodeInputExpression, inputMapping: Map[String, OutputPort]): ErrorOr[ConditionalNodeWithInputs] = {
+    val graphNodeSetter = new GraphNode.GraphNodeSetter()
+    val inputPortLinker = GraphNode.linkInputPort("", inputMapping, graphNodeSetter.get) _
+
+    val conditionalExpressionType: ErrorOr[Unit] = condition.evaluateType flatMap {
+      case WdlBooleanType => Valid(())
+      case other => s"Cannot base a conditional on non-boolean type ${other.toWdlString}".invalidNel
+    }
+
+    val conditionalExpressionValidation: ErrorOr[InstantiatedExpression] = condition.instantiateExpression(graphNodeSetter)
+
+    val linkedInputPortsAndGraphInputNodes: Set[LinkedInputPort] = innerGraph.nodes.inputDefinitions.map(inputPortLinker)
+    val linkedInputPorts = linkedInputPortsAndGraphInputNodes.map(_.newInputPort)
```

Comment:
I'm having trouble following these 3 lines of code, perhaps being explicit about types and maybe an overview of the goal?